flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/FlinkKubeClient.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.kubeclient;
+
+import org.apache.flink.client.deployment.ClusterSpecification;
+import org.apache.flink.kubernetes.kubeclient.resources.FlinkPod;
+import org.apache.flink.kubernetes.kubeclient.resources.FlinkService;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * The client to talk with kubernetes.
+ */
+public interface FlinkKubeClient extends AutoCloseable {
+
+	/**
+	 * Create kubernetes config map, include flink-conf.yaml, log4j.properties.
+	 */
+	void createConfigMap() throws Exception;
```

Comment:
I notice that some of "create" functions are blocking while others are non-blocking. What is the different between them and how is the decision made?

Reply:
All the create operation is asynchronous, including `createConfigMap createInternalService createFlinkMasterDeployment createTaskManagerPod`.  We just add a `CompletableFuture` for creating service because we need to make sure the service are created successfully. Then we could get the uuid of internal service and use the rest service to submit job.